automata/pda/stack.py

```diff
@@ -2,19 +2,20 @@
 """Classes and methods for working with PDA stacks."""
 from __future__ import annotations
 
-import collections
+from dataclasses import dataclass
 from typing import Iterator, Sequence, Tuple
 
 
-class PDAStack(collections.namedtuple("PDAStack", ["stack"])):
+@dataclass(frozen=True)
+class PDAStack:
     """A PDA stack."""
 
+    __slots__ = ("stack",)
+
     stack: Tuple[str, ...]
 
-    # TODO can we get rid of this? Kinda weird inheritance here
-    def __new__(cls, elements: Sequence[str]):
-        """Create the new PDA stack."""
-        return super(PDAStack, cls).__new__(cls, tuple(elements))
+    def __init__(self, elements: Sequence[str]):
```

Comment:
I think annotating this as a sequence and conversion to a tuple makes the most sense (better than what was done before).

Reply:
@eliotwrobson Me too. I guess it's a benefit of no longer inheriting from `namedtuple` (although my thought was that if we inherit from `Tuple`, the conversion happens automatically and we no longer need any explicit `__init__` or `__new__` definition whatsoever).

Reply:
I think that might make sense, because the internal functions all treat this class as immutable.

Reply:
@eliotwrobson So ironically, I actually just thought of an argument NOT to inherit from tuple:

If we ever decide in the future that `PDAStack` should have any additional data members, we would need to resort back to adding a `__new__` definition (with `super()`) because `tuple()` only accepts a single argument natively. Maybe that's not an issue now while `PDAStack` still has a single data member `stack`, but trying to consider a solution that is future-proof.

I was also trying to think about consistency with similar sequence-type structures like `TMTape`, which accepts a `tape` parameter as input. So if we make `PDAStack` inherit from `Tuple`, would it suggest we should make `TMTape` inherit from `tuple` as well? But `TMTape` requires multiple arguments, so we'd have to use `__new__` in that case.

Maybe I'm just overthinking this in some futile pursuit of API perfection. I suppose it really shouldn't matter much, since `PDAStack` and friends aren't publicly-documented classes (beyond being part of return values, anyway). But gosh, I just feel so indecisive about this now lol. I feel like I could go either way.

Reply:
@calebevans Ahh I see what you mean. I lean towards consistency with the TMTape and not having to define the __new__ function, since I'm generally a fan of doing as few things under the hood as possible as to make code more readable and easier to debug.

Reply:
@eliotwrobson wrong caleb

Reply:
@eliotwrobson This is why it's so nice to have you on the project now. 😌

You have tipped the scale for me, and I am in agreement to avoid inheriting from native types here (and perhaps as a rule-of-thumb going forward, as well).

And I guess the best part of this that by just leaving the object structure as-is, we are not changing the public (or private) API in this respect. One less thing to document lol